[material-ui][docs] Revise the Button page (#40446)
mui-org/material-ui: docs/data/material/components/buttons/buttons.md

Icons are also appropriate for toggle buttons that allow a single choice to be selected or
deselected, such as adding or removing a star to an item.
The Material UI Icon Button component accepts the same props and values as the Button for [sizes](#sizes) and [colors](#colors).

{{"demo": "IconButtons.js"}}

### Sizes

For larger or smaller icon buttons, use the `size` prop.

{{"demo": "IconButtonSizes.js"}}
### Handling clicks

### Colors

Use `color` prop to apply theme color palette to component.

{{"demo": "IconButtonColors.js"}}

## File upload
Material UI Buttons accept an `onClick` handler that's applied to the root DOM element.

To create a file upload button, turn the button into a label using `component="label"` and then create a visually-hidden input with type `file`.

{{"demo": "InputFileUpload.js"}}
```jsx
<Button
  onClick={() => {
    alert('clicked');
  }}
>
  Click me
</Button>
```

## Customization

Here are some examples of customizing the component.
You can learn more about this in the [overrides documentation page](/material-ui/customization/how-to-customize/).

> {{"demo": "CustomizedButtons.js", "defaultCodeOpen": false}}

Reviewer: This one would be good to maintain in the Customization section, but maybe with more drastically visually-looking designs? I can help with that!

Reviewer: Or not, I'm not sure, given it's just a simple visual customization without any complex/sneaky trick. What do you think? If we're not keeping it, let's remove the file!

Reviewer: Awesome! I'll work on adding more later on — Material UI as a whole is lacking more entirely customized examples!

# Button

<p class="description">Buttons enable users to take actions and make choices.</p>

{{"component": "modules/components/ComponentLinkHeader.js"}}

## Introduction

Buttons communicate actions that users can take.
The Material UI Button component replaces the native HTML `<button>` element and offers expanded options for styling and accessibility.

{{"demo": "IntroButtons.js"}}

> ## Basics

Reviewer: ```suggestion
## Basics

### Import
```

Should we add it? We did it on the Snackbar page, but I guess we never formalized it as part of the template!

Reviewer: Uhm, that makes sense, yeah! Well, I have no strong preferences either, so I'm happy to go without it, too.

Reviewer: Yeah, these first two Base UI results, for example, are from the Components API and Hooks API tabs, respectively, which makes me think we should probably include them somehow in the results (so you know what that's about).

Reviewer 2: The problem here to me is how we rank Algolia results. If I'm on a Material UI page, no way I want to see Base UI or MUI System results in between. I want first as section that is all about the product I'm looking at, and second an area for other results.

Reviewer: I think that's _another_ problem on top of the one Sam shared. Imagine if Sam's screenshot is a search result when browsing through the Base UI docs. The results order is correct, but still, there's an opportunity to improve how we differentiate pages or sections with the same heading text.

```jsx
import Button from '@mui/material/Button';
```

### Links

The Material UI Button accepts an `href` prop, which turns the HTML `<button>` element into an `<a>` with the target link.

```jsx

> <Button href="/home">Home</Button>

Reviewer 2: [#suggestion](https://www.notion.so/mui-org/Olivier-Tassinari-a6d0f1d19f8a4cfbb20469878ded8aa8?pvs=4#ac52aa0d7db5447b832416ef2c790f83)

### Colors

Every palette included in the theme is available via the `color` prop.

{{"demo": "ColorButtons.js"}}

:::success
If the default colors aren't sufficient, you can add custom colors and remove defaults that you don't need.
See the [Palette—Custom colors](/material-ui/customization/palette/#custom-colors) doc for more info.
:::

### Sizes

The Button component comes in three sizes: `small`, `medium` (default), and `large`.

{{"demo": "ButtonSizes.js"}}

#### Width

Use the `fullWidth` prop to stretch the Button to fill its container.

{{"demo": "FullWidthButton.js"}}

### Start and end icons

Use the `startIcon` or `endIcon` prop to append an icon to the beginning or end of a Button's contents.

{{"demo": "IconLabelButtons.js"}}

### Ripple effect

The ripple effect provides visual feedback to the user that the Button is being clicked or focused.
Use the `disableFocusRipple` prop to remove only the focus effect, or `disableRipple` to remove the ripple when focusing and clicking.

:::warning
These props remove _all_ default styles from the `:focus-visible` state, which means the user won't have any visual indication that the Button is focused.

> You can target the `.Mui-focusVisible` class to define your own styles for a better user experience.

Reviewer: ```suggestion
You can target the `.Mui-focusVisible` class to define your own styles for an accessible user experience.
```

Thought of calling out the accessibility impact of having properly styled focus styles here instead of saying "better"; maybe that's more specific... and better? 😅

> 

Reviewer 2: Would it make more sense? We already know it's a button page, and "Material" doesn't exists by itself.

"Material 3 Button" -> "Material Design v3"

Reviewer: > ...and "Material" doesn't exists by itself

The official Material Design docs do have plenty of mentions of "Material 3", using "Material" standalone like that. [See the homepage](https://m3.material.io/) — "M3", "Material talks", "Material blog", etc.

Reviewer 2: But is it more as a shorthand? Using the full version could avoid confusing. Is Material v3 the same thing as Material UI v3?

Reviewer: Heh, the "Is Material as in Material Design or as in Material UI?" question is _the_ question 😬 Anyway, no hard preference on using Material Design v3 — my take was to have a matching terminology with the official MD docs, but it may be worth the clarification.

Reviewer: Oh, very nice link (it talks about the default color, too, huh!). I'd say that using "Material Design" to differentiate from Material UI is probably the safest approach most of the time, not just when we initially refer to it.

{{"demo": "ButtonMaterialYouPlayground.js", "hideToolbar": true, "bg": "playground"}}

To learn more about Material UI's MD3 implementation, visit the [Material 3 Components documentation](/material-ui/guides/material-3-components/).

## Anatomy

The Material UI Button component is composed of a single root `<button>` element with a `<span>` nested inside that houses the touch ripple effect.
Optional [start and end icons](#start-and-end-icons) are rendered inside of `<span>` elements.

```html
<button class="MuiButtonBase-root MuiButton-root">
  <span class="MuiButton-startIcon">
    <svg class="MuiSvgIcon-root"></svg>
  </span>
  <!-- Button contents here -->
  <span class="MuiButton-endIcon">
    <svg class="MuiSvgIcon-root"></svg>
  </span>
  <span class="MuiTouchRipple-root"></span>
</button>

> ```

Reviewer 2: I think we should reverse the links, I tend to click on Base UI thinking it's the Button unstyled, but it's the homepage.
 
![Screenshot 2024-01-30 at 18 16 22](https://github.com/mui/material-ui/assets/3165635/4d3d7354-5480-4e47-95c8-d6cca17fc248)

Reviewer: Or remove the "Base UI" link entirely?!

Reviewer 2: Alright, done #40875.
